### PR TITLE
refactor(core): typed ConflictError + document MemoryError variants (#115)

### LINKED ISSUE #115: [super-qa] medium: MemoryError variants are stringly-typed catch-alls
## Finding: Stringly-typed Error Variants

- **ID:** M14
- **Severity:** medium
- **Category:** design
- **Location:** src/error.rs:1-42
- **Detected by:** code-reviewer agent

## Description

6 MemoryError variants (Conflict, Migration, Internal, Pool, Bootstrap, Reranker) are all \`String\`-wrapped catch-alls. They're used for fundamentally different error conditions. Callers cannot pattern-match on the specific failure without string parsing.

## Suggested Approach

Split into specific error types. Start with Migration and Conflict as they're the most overloaded:
- \`Migration { MissingUpcaster, IncompatibleSchema, ... }\`
- \`Conflict { InvalidPolicy, MutuallyExclusive, ScopeValidation, ... }\`

### LINKED ISSUE #496: [super-qa] core: low/info findings (55)
> ⚠️ Agent analysis; not source-verified. Grouped low/info for **area:core**.

Grouped low/info findings for area **core** (55):

- **low** `src/async_engine.rs:41-43` — tokio::task::JoinError silently mapped to MemoryError::Pool (connection-pool variant) (`core-root/correctness-join-error-mapped-to-pool-variant`)
- **low** `src/types.rs:17,33-41` — Outcome Display produces lowercase but serde JSON and payload doc specify PascalCase (`core-root/correctness-outcome-display-vs-serde-casing-mismatch`)
- **low** `src/resume/context.rs:33` — ResumeConfig::Default captures Utc::now() at construction time — stale if config is stored before use (`resume/correctness-stale-now-in-default`)
- **low** `src/scope/tree.rs:61-69` — ancestors() called with an unknown scope_id returns [scope_id] instead of empty (`scope/correctness-ancestors-unknown-id`)
- **low** `src/bootstrap/metrics.rs:11, src/bootstrap/mod.rs:162` — `max_turns: usize` uses `0` as a sentinel for "no limit" — stringly-typed boolean embedded in an integer (`bootstrap/design-sentinel-zero-max-turns`)
- **low** `src/bootstrap/extract.rs:79` — `format!("{:?}", episode.category)` embeds Rust Debug repr into stored JSON metadata (`bootstrap/design-debug-format-for-category-in-metadata`)
- **low** `src/types.rs:150,336` — `Edge.relation_type` and `NewEdge.relation_type` are untyped strings with a closed value set in practice (`core-root/design-stringly-typed-relation-type`)
- **low** `src/traits.rs:43-57` — `SummaryGenerator` duplicates `EmbeddingProvider::embed` instead of composing (`core-root/design-summary-generator-embed-duplication`)
- **low** `memory-engine-embed/src/http.rs:13,22-25` — `endpoint` stored as raw `String` — invalid URLs fail only at first network call (`embed/design-stringly-typed-endpoint`)
- **low** `src/engine/mod.rs:59-74` — EngineConfig has all-pub fields — constructor invariants can be bypassed (`engine/design-engine-config-pub-fields`)
- **low** `src/engine/outcome.rs:91-93` — Outcome counts matched against literal strings — brittle to serde rename on the Outcome enum (`engine/style-outcome-stringly-typed-match`)
- **low** `src/bootstrap/extract.rs:60-96` — `KeywordExtractor::extract` silently omits `session_outcome` from metadata for `Indeterminate` sessions (`bootstrap/documentation-keyword-extractor-indeterminate-metadata`)
- **low** `src/bootstrap/outcome.rs:30` — `OutcomeSignals::final_user_sentiment` typed `Option<Sentiment>` but is always `Some` in `classify_outcome` (`bootstrap/documentation-outcome-signals-option-sentiment`)
- **low** `src/types.rs:256-257` — AddFactOptions::importance doc promises [0, 1] but no validation is enforced (`core-root/documentation-add-fact-options-importance-unenforced`)
- **low** `src/error.rs:3-54` — Most MemoryError enum variants have no doc comment; Conflict is semantically overloaded for parameter validation without documentation (`core-root/documentation-error-variants-undocumented`)
- **low** `/home/mroynard/dev/memory-engine/memory-engine-embed/src/http.rs:44-45,89` — Private helper docs missing # Errors section despite returning Result (`embed/documentation-private-methods-missing-errors-section`)
- **low** `src/engine/mod.rs:126-133` — `open()` doc incorrectly states it writes `embed_dim` on first open even in read-only mode (`engine/documentation-open-readonly-first-open`)
- **low** `src/engine/resume.rs:17; src/engine/graph.rs:16-19; src/engine/inspect.rs:52` — Three stale Phase-5 forward-reference comments in public API docs (`engine/documentation-stale-phase5-markers`)
- **low** `src/resume/context.rs:62,46` — "all pinned facts" claim ignores `pinned_cap` truncation (`resume/documentation-pinned-all-misleading`)
- **low** `src/scope/tree.rs:60-69` — ancestors silently returns empty Vec for unknown scope_id — not documented (`scope/documentation-ancestors-unknown-id`)
- **low** `src/store/scopes.rs:11` — #[allow(dead_code)] on entire ScopeStore impl — accumulating API surface with no callers (`scope/stale-dead-code-allow-scope-store`)
- **low** `src/bootstrap/filter.rs:322-324` — keyword_prefilter allocates three lowercased strings per turn where two suffice (`bootstrap/performance-redundant-string-allocations-prefilter`)
- **low** `src/async_engine.rs:309-312,350-353` — Ref parameters immediately cloned to satisfy spawn_blocking 'static bound (`core-root/performance-forced-clone-ref-params-spawn-blocking`)
- **low** `memory-engine-embed/src/http.rs:66,105,148,152,155` — Unnecessary deep clone of `serde_json::Value` per embedding vector when parsing (`embed/performance-value-clone-per-embedding`)
- **low** `src/bootstrap/filter.rs:322-324, 389-412` — Stderr iterated twice per turn in `keyword_prefilter`: once in `build_searchable_text`, once in `collect_stderr` (`bootstrap/refactoring-redundant-stderr-iteration`)
- **low** `src/bootstrap/filter.rs:344-383` — Three structurally identical `CandidateEpisode` push blocks in `keyword_prefilter` — extract data table (`bootstrap/refactoring-keyword-prefilter-repetition`)
- **low** `src/bootstrap/metrics.rs:44-73` — `BootstrapReport::merge` is a 30-line manual field-by-field accumulator — fragile under new field addition (`bootstrap/refactoring-report-merge-manual-fields`)
- **low** `src/engine/bootstrap.rs:31-41, src/engine/bootstrap.rs:71-81` — Scope resolution block copy-pasted identically in `bootstrap_session` and `bootstrap_directory` (`bootstrap/refactoring-scope-resolution-duplication`)
- **low** `src/pool/connection_pool.rs:23-24, 229-231` — embed_dim stored in ConnectionPool with #[allow(dead_code)] — pool owns data it does not use (`pool/refactoring-embed-dim-responsibility`)
- **low** `src/resume/context.rs:69-113, src/store/facts.rs:837-839` — resume_context fully deserializes embedding BLOBs for every returned fact even though the resume path does not use embeddings (`resume/refactoring-embedding-deserialization-cost`)
- **low** `src/types.rs:95,136,259,376,493,532,564 (serde_json::Value fields)` — serde_json::Value payload/metadata fields are an under-bounded trust boundary (CWE-502/CWE-20/CWE-400) (`core-root/deserialization-value-trust-boundary`)
- **low** `src/types.rs:126,162,164,227,234,244,489,497 (Vec<f32>/Vec<i64> fields)` — Embedding and id-list Vec fields deserialize without a length cap (CWE-400/CWE-770) (`core-root/deserialization-unbounded-vec`)
- **low** `memory-engine-embed/src/http.rs:64-71,148,155,167,234` — Embedding array deserialized into Vec<f32> before expected_dim validation (`embed/dos-vecf32-alloc-before-dim-check`)
- **low** `src/pool/connection_pool.rs:194-203` — read() blocks indefinitely with no timeout; a leaked ReadGuard deadlocks all readers (CWE-400) (`pool/resource-exhaustion-untimed-blocking-read`)
- **low** `memory-engine-embed/src/http.rs:22-40` — Five-parameter positional constructor is prone to argument-order mistakes (`embed/style-positional-constructor`)
- **low** `src/pool/connection_pool.rs:94, 100, 163, 168` — path.to_string_lossy() silently corrupts non-UTF-8 paths when opening connections (`pool/style-lossy-path-conversion`)
- **low** `src/bootstrap/metrics.rs:44-73` — Property-based test missing for `BootstrapReport::merge` weighted-average invariant (`bootstrap/testing-proptest-report-merge-identity`)
- **low** `src/async_engine.rs:28-34` — Only doc-test in scope is marked rust,ignore — zero runnable examples across public API (`core-root/testing-doc-test-ignore`)
- **low** `memory-engine-embed/src/http.rs:193-195` — Bearer auth is only tested for embed(), not for embed_batch() (`embed/testing-batch-auth-uncovered`)
- **low** `src/engine/outcome.rs:94` — Forward-compat `_ => {}` branch in `get_outcome_counts` is never exercised (`engine/untested-outcome-unknown-variant`)
- **low** `src/engine/mod.rs (pub fn open, open_memory, add_fact, query, consolidate, forget)` — Major public API methods have no runnable `# Examples` doc-tests (`engine/doctest-missing-public-api`)
- **low** `src/pool/connection_pool.rs:88-115,187-202` — open() with read_pool_size=0 on a file-backed path is untested and silently routes reads through write_conn (`pool/testing-read-pool-size-zero-file-backed`)
- **low** `src/pool/connection_pool.rs:246-365` — No proptest coverage for pool constructor parameter invariants (`pool/testing-no-property-based-tests`)
- **low** `src/pool/connection_pool.rs:151-181` — No fuzz targets exist in the workspace for pool's untrusted-path inputs (`pool/testing-no-fuzz-targets`)
- **low** `tests/` — Pool is never tested as an external consumer — all integration tests go through MemoryEngine (`pool/testing-no-integration-test-for-pool-directly`)
- **low** `src/resume/context.rs:80-102` — Cap fields never tested at boundary values 0 or 1 (`resume/testing-cap-boundary-zero`)
- **low** `src/scope/tree.rs:313-318 (proptest_scope / scope_segment / scope_path)` — Proptest strategies cover only colon-containing segments and at most 3-depth paths (`scope/testing-proptest-strategy-restricted-alphabet`)
- **low** `src/scope/tree.rs:9-168 (all pub fn doc comments)` — No public item in src/scope/ has a runnable doc-test example (`scope/testing-no-doctests-on-public-api`)
- **info** `src/bootstrap/metrics.rs:93` — `PrewarmMetrics` doc cites opaque acronyms `R3` and `APC` with no expansion (`bootstrap/documentation-prewarm-metrics-opaque-acronyms`)
- **info** `src/bootstrap/filter.rs:1, src/bootstrap/filter.rs:313` — "NeuroStack" used as a proper noun in public-facing doc comments without definition or link (`bootstrap/documentation-neurostack-unexplained-term`)
- **info** `src/pool/connection_pool.rs:23-24,228-229` — `#[allow(dead_code)]` on `embed_dim` field and getter references closed issue #108 (`pool/documentation-dead-code-allow-stale-comment`)
- **info** `src/scope/tree.rs:313-318` — Proptest scope_segment generates only colon-namespaced labels — plain-word labels never exercised (`scope/documentation-proptest-segment-coverage-gap`)
- **info** `memory-engine-embed/src/http.rs:221-225` — Manual body truncation (1000-char limit) is a hand-rolled snippet with no test (`embed/refactoring-truncate-helper`)
- **info** `src/types.rs:100-103,108-109,140-141,217-218` — serde(default) fields reviewed: defaults are benign, no value-injection risk (CWE-502) (`core-root/serde-default-injection-audit`)
- **info** `src/bootstrap/mod.rs:158` — `OutcomeSignals` always discarded in the production pipeline — useful only in tests (`bootstrap/style-outcome-signals-unused-in-pipeline`)

### PR DESCRIPTION
**Pull request by dutiona**
## Summary

Re-scoped slice of #115: split the stringly-typed `MemoryError::Conflict(String)` into a typed `ConflictError` sub-enum, and document **every** `MemoryError` variant.

## `ConflictError` variants (derived from the 26 real construction sites — not invented; no catch-all `Other`)

| Variant | Meaning | Sites |
| --- | --- | --- |
| `QueryValidation(String)` | Incompatible/incomplete `MemoryQuery` options (half-open period, `valid_at` + period, `SearchMode` missing its input) | 5 — `engine/query.rs` |
| `PolicyParameter(String)` | Config/policy parameter out of range (half-life, weights, ratios, percentiles) | 6 — `ForgetPolicy::validate` (4) + `DreamCycleConfig::validate` (2) |
| `ScopeLabel(String)` | Malformed scope label/path segment (empty, contains `/`, too long, whitespace) | 5 — `store/scopes.rs` |
| `TargetExists` *(unit)* | Restore target DB path already exists | 2 — `engine/restore.rs` |
| `TargetNotEmpty` *(unit)* | Restore into a non-empty DB | 1 — `inspect/restore.rs` |
| `DumpTargetIsLiveDatabase` *(unit)* | Dump path resolves to the live DB file | 1 — `inspect/dump.rs` |
| `Arbitration(String)` | A consumer-supplied trait reported a failure | 2 — `ConflictArbiter`/`EmbeddingProvider` test mocks |

**26 sites accounted for**: 22 constructions migrated to typed variants + 3 `matches!(_, MemoryError::Conflict(_))` assertions (unchanged, wildcard binds the inner enum) + 1 MCP `to_mcp_error` match arm (unchanged; `ConflictError: Display` keeps `"conflict: <msg>"` output). No `Conflict(format!/String)` remains.

## Design

- `Conflict(String)` → `Conflict(#[from] ConflictError)`; outer `#[error("conflict: {0}")]` delegates to `ConflictError`'s Display, so all user-facing messages are **preserved** (string variants verbatim; unit variants carry their text in `#[error(...)]`).
- `ConflictError` derives `Debug` + `thiserror::Error`, is `#[non_exhaustive]`, and is re-exported at the crate root via `pub use error::*` (`memory_engine::ConflictError`).
- Every `MemoryError` variant now has a `///` doc stating when it is returned (the enum was already `#[non_exhaustive]`; most variants were undocumented).

## Verification (workspace gate — 4 crates: core, cli, mcp, embed)

- `cargo build --workspace` ✅ (and `--all-features` ✅)
- `cargo test --workspace` ✅ — 925 passed, 6 ignored
- `cargo clippy --workspace --all-targets` ✅ — exit 0, **zero new** warnings (pre-existing ~#539 pedantic/nursery only)
- `cargo fmt --check` ✅

Fixes #115
Resolves #496 (core-root/documentation-error-variants-undocumented).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by dutiona** [comment]
## Adversarial review summary (replaces /super-review)

Reviewed by **3 clean-slate adversarial reviewers** (error-design lens pack):

| Lens                | Verdict    | Key finding                                                                                                                                                                                                                                                                                                                                      |
| ------------------- | ---------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| **correctness**     | ✅ APPROVE | All 22 construction sites migrated to typed `ConflictError`; branch-wide grep confirms **no `Conflict(String)`/`format!` remains**. Messages **byte-preserved** (`#[error("{0}")]` passthrough + verbatim unit-variant text); MCP `conflict: <msg>` output unchanged. Pure type change — 925 tests pass, `matches!` assertions already wildcard. |
| **api-design**      | ✅ APPROVE | All 7 variants **derived from the 26 real sites** (no invented categories, no catch-all `Other`). `thiserror` Display + `#[from]` + `#[non_exhaustive]` sound. All 16 `MemoryError` + 7 `ConflictError` variants now documented (**resolves #496**).                                                                                             |
| **maintainability** | ✅ APPROVE | Zero `Conflict(String)`/`format!` anywhere (core/cli/mcp/embed). 22 migrated + 3 wildcard `matches!` + 1 MCP `Display` arm = **26 accounted**. Split correctly scoped to `Conflict`; other `String` variants documented but untouched.                                                                                                           |

**Result: 3/3 approve, 0 blockers.** `ConflictError` variants: `QueryValidation`, `PolicyParameter`, `ScopeLabel`, `TargetExists`, `TargetNotEmpty`, `DumpTargetIsLiveDatabase`, `Arbitration`.

Nits (non-blocking): `scopes.rs` empty-path → `ScopeLabel` (doc-covered, message preserved; branch is pre-existing dead code); an `EmbeddingProvider` test mock uses `ConflictError::Arbitration` (doc covers "or other injected provider").

The remaining `String` variants (`Migration`/`Internal`/`Archive`/…) are tracked toward a full typed hierarchy in the follow-up issue.

File: src/store/scopes.rs [hunk 1]
```diff
@@ -107,7 +107,9 @@ impl<'a> ScopeStore<'a> {
     pub fn ensure_path(&self, path: &str) -> Result<i64> {
         let segments: Vec<&str> = path.split('/').collect();
         if segments.is_empty() {
-            return Err(MemoryError::Conflict("scope path must not be empty".into()));
+            return Err(MemoryError::Conflict(ConflictError::ScopeLabel(
+                "scope path must not be empty".into(),
+            )));
         }
```
[the remark this answers is not in the document]
Comment (dutiona) [reply src/store/scopes.rs]: Addressed in ef5919c — replaced the dead `segments.is_empty()` check with an up-front `path.is_empty()` guard (str::split always yields >=1 element), avoiding the allocation and returning the path-specific "scope path must not be empty" message via `ConflictError::ScopeLabel`. Re-gate green: 925 tests pass. Resolving.
